README.md
# WPGitHubAPI
`[wga]` Shortcode to obtain resources from official GitHub API v3.

NOTE: WPGitHubAPI plugin uses transients to cache the counts for `24 * HOUR_IN_SECONDS`.

## Shortcode
- `[wga]` shortcode accepts following parameters
    + `u='api_url'` where `api_url` is the part of API after `https://api.github.com/` E.g. `users/ahmadawais` will return my [profile data](https://api.github.com/users/ahmadawais).
    + `d='data'` where `data` is the object's data you want to retrieve E.g. `followers` will return the number of followers for my profile.
    + GitHub Readme Parser with `r='y'` attribute in the shortcode.
    + Remove `h1` with `h1=y|n` attribute in the shortcode.
    + Convert `h1` to `h2` with `h2=y|n` attribute in the shortcode.

### Examples
- My GitHub Followers Count: `[wga u='users/ahmadawais' d='followers']`
- For more examples read [GitHub API Documentation](https://developer.github.com/v3/) or visit [GitHub API](https://api.github.com/).

### Screenshot
![](https://i.imgur.com/jraBp7C.png)

## To do 
- [x] WGA shortcode
- [ ] WGA Documentation

## Updates
You can use [GitHub Updater](https://github.com/afragen/github-updater) to receive updates for this plugin. 

## Changelog

### Version 0.0.8
- IMPROVE: Gracefully show the GitHub repo link on API rate limiting.

### Version 0.0.7
- IMPROVE: PHP 7 compat — thanks to @[tfirdaus](https://github.com/tfirdaus) for PR #[1](https://github.com/ahmadawais/WPGitHubAPI/pull/1).
- FIX: PHP CS Issues like weird spacing and tabs.
- NEW: Transients below 172 characters.

### Version 0.0.6
- NEW: Should there be `h1` with `h1=y|n`.
- NEW: Should convert `h1` to `h2` with `h2=y|n`.

### Version 0.0.5
- NEW: Basic PHPUnit tests.
- NEW: GitHub Readme Parser with `r='y'` attribute in the shortcode.

### Version 0.0.4
- WP.org Repo submission.
- NEW: `README.txt`.

### Version 0.0.3
- NEW: Custom transients names with API_URL and Data.
- NEW: Regex Remove `/` in the beginning or at the end.

### Version 0.0.2
- FIX: Custom data related transients.

### Version 0.0.1
- First beta version.

## License
Licensed under GNU GPL v2.0. Copyright (C) 2016 [Ahmad Awais](http://AhmadAwais.com/).

---
### 🙌 [WPCOUPLE PARTNERS](https://WPCouple.com/partners):
This open source project is maintained by the help of awesome businesses listed below. What? [Read more about it →](https://WPCouple.com/partners)

<table width='100%'>
    <tr>
        <td width='225'><a target='_blank' href='https://www.gravityforms.com/?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mtrE/c' /></a></td>
        <td width='225'><a target='_blank' href='https://kinsta.com/?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mu5O/c' /></a></td>
        <td width='225'><a target='_blank' href='https://wpengine.com/?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mto3/c' /></a></td>
        <td width='225'><a target='_blank' href='https://www.sitelock.com/?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mtyZ/c' /></a></td>
    </tr>
    <tr>
        <td width='225'><a target='_blank' href='https://wp-rocket.me/?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mtrv/c' /></a></td>
        <td width='225'><a target='_blank' href='https://blogvault.net/?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mtph/c' /></a></td>
        <td width='225'><a target='_blank' href='http://cridio.com/?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mtmy/c' /></a></td>
        <td width='225'><a target='_blank' href='http://wecobble.com/?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mtrW/c' /></a></td>
    </tr>
    <tr>
        <td width='225'><a target='_blank' href='https://www.cloudways.com/?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mu0C/c' /></a></td>
        <td width='225'><a target='_blank' href='https://www.cozmoslabs.com/?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mu9W/c' /></a></td>
        <td width='225'><a target='_blank' href='https://wpgeodirectory.com/?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mtwv/c' /></a></td>
        <td width='225'><a target='_blank' href='https://www.wpsecurityauditlog.com/?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mtkh/c' /></a></td>
    </tr>
    <tr>
        <td width='225'><a target='_blank' href='https://www.liquidweb.com//?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mtnt/c' /></a></td>
        <td width='225'><a target='_blank' href='https://WPCouple.com/contact?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mu3F/c' /></a></td>
        <td width='225'><a target='_blank' href='https://WPCouple.com/contact?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mu3F/c' /></a></td>
        <td width='225'><a target='_blank' href='https://WPCouple.com/contact?utm_source=WPCouple&utm_medium=Partner'><img src='http://on.ahmda.ws/mu3F/c' /></a></td>
    </tr>
</table>



CHANGELOG.md
# WPGitHubAPI

### Version 0.0.8
- IMPROVE: Gracefully show the GitHub repo link on API rate limiting.
### Version 0.0.7
- IMPROVE: PHP 7 compat — thanks to @[tfirdaus](https://github.com/tfirdaus) for PR #[1](https://github.com/ahmadawais/WPGitHubAPI/pull/1).
- FIX: PHP CS Issues like weird spacing and tabs.
- NEW: Transients below 172 characters.

### Version 0.0.6
- NEW: Should there be `h1` with `h1=y|n`.
- NEW: Should convert `h1` to `h2` with `h2=y|n`.

### Version 0.0.5
- NEW: Basic PHPUnit tests.
- NEW: GitHub Readme Parser with `r='y'` attribute in the shortcode.


### Version 0.0.4
- WP.org Repo submission.
- NEW: `README.txt`.

### Version 0.0.3
- NEW: Custom transients names with API_URL and Data.
- NEW: Regex Remove `/` in the beginning or at the end.

### Version 0.0.2
- FIX: Custom data related transients.

### Version 0.0.1
- First beta version.


tests/bootstrap.php
<?php

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

require_once $_tests_dir . '/includes/functions.php';

function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/WPGitHubAPI.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

require $_tests_dir . '/includes/bootstrap.php';


wpgithubapi.php
<?php
/**
 * Plugin Name: WP GitHub API
 * Plugin URI: https://github.com/ahmadawais/WPGitHubAPI
 * Description: Shortcode to obtain resources from official GitHub API v3.
 * Author: mrahmadawais, WPTie
 * Author URI: http://AhmadAwais.com/
 * Version: 0.0.8
 * License: GPLv2 or later
 * License URI: http://www.gnu.org/licenses/gpl-2.0.txt
 * GitHub Plugin URI: https://github.com/ahmadawais/WPGitHubAPI/
 *
 * @package WGA
 */

// Exit if accessed directly.
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

/**
 * Define global constants.
 *
 * @since 0.0.2
 */
if ( ! defined( 'WGA_NAME' ) ) {
	define( 'WGA_NAME', trim( dirname( plugin_basename( __FILE__ ) ), '/' ) );
}

if ( ! defined( 'WGA_DIR' ) ) {
	define( 'WGA_DIR', WP_PLUGIN_DIR . '/' . WGA_NAME );
}

if ( ! defined( 'WGA_URL' ) ) {
	define( 'WGA_URL', WP_PLUGIN_URL . '/' . WGA_NAME );
}

/**
 * Class: WP_GitHub_API_Shortcodes.
 *
 * @since 1.0.0
 */
if ( file_exists( WGA_DIR . '/class/class-wp-github-api-shortcodes.php' ) ) {
	require_once( WGA_DIR . '/class/class-wp-github-api-shortcodes.php' );
}

new WP_GitHub_API_Shortcodes();


class/class-wp-github-api-shortcodes.php
<?php
/**
 * Class: WP_GitHub_API_Shortcodes
 *
 * Main class for shortcodes.
 *
 * @since    0.0.1
 * @package WGA
 */

// Exit if accessed directly.
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

/**
 * Parsedown.
 *
 * @since 0.0.5
 * @link https://github.com/erusev/parsedown
 */
if ( file_exists( WGA_DIR . '/class/class-parsedown.php' ) ) {
	require_once( WGA_DIR . '/class/class-parsedown.php' );
}

if ( ! class_exists( 'WP_GitHub_API_Shortcodes' ) ) :

	/**
	 * WP_GitHub_API_Shortcodes.
	 *
	 * @since 0.0.1
	 */
	class WP_GitHub_API_Shortcodes {

		/**
		 * Construct.
		 *
		 * @since 0.0.1
		 */
		public function __construct() {
			// Load the plugin.
			add_action( 'plugins_loaded', array( $this, 'register' ) );
		}

		/**
		 * Register.
		 *
		 * @since 0.0.1
		 */
		public function register() {
			// Add the shortcode.
			add_shortcode( 'wga', array( $this, 'shortcode' ) );
		}

		/**
		 * Shortcode.
		 *
		 * @param Attributes $atts API URL|Data.
		 * @since 0.0.1
		 */
		public function shortcode( $atts ) {
			// Save $atts.
			$_atts = shortcode_atts( array(
				'u'  => 'repos/ahmadawais/WPGulp', // API URL.
				'd'  => 'name', // Data.
				'r'  => 'n', // Readme.
				'h1' => 'y', // Remove h1.
				'h2' => 'y', // Convert h1 to h2.
			), $atts );
			// API URL.
			$_api_url = $_atts['u'];

			// Remove `/` in the beginning or at the end.
			$_api_url = preg_replace( '/^\/|\/$/', '', $_api_url );

			// Data.
			$_data = $_atts['d'];

			// Is Readme?
			$_readme = $_atts['r'];

			// Is h1 in content? Mostly the title.
			$_is_h1 = $_atts['h1'];

			// Is Convert h1 to h2?
			$_ish1toh2 = $_atts['h2'];

			// Get WPTAlly's object.
			$wpgapi_response = $this->get_api( $_api_url, $_data );

			// Bail if there is an error.
			if ( ! is_object( $wpgapi_response ) || is_wp_error( $wpgapi_response ) ) {
				return $wpgapi_response;
			}

			// Decode b64 if it is Readme to HTML.
			if ( 'y' == $_readme ) {
				// Get the response.
				$_readme_b64 = $wpgapi_response->$_data;

				// Nothingness. API Rate Limit maybe! Gracefully show the GitHub repo link.
				if ( $_readme_b64 === NULL ) {
					// Remove repo from it.
					$get_user_repo = preg_replace( '/\brepos\b/i', '', $_api_url );
					$get_user_repo = preg_replace( '/\breadme\b/i', '', $get_user_repo );

					$github_url = 'https://github.com' . $get_user_repo;

					// Debug.
					if ( true === WP_DEBUG ) {
						$api_message = $wpgapi_response->message;
						return 'Content not available right now... <br> <div><strong>REASON: </strong>' . $api_message . '</div><br> <div>VAR DUMPED: </div>' . var_dump( $wpgapi_response );
					}
					return 'How about you checkout the <a href="' . $github_url . '">repository on GitHub</a> →';
				}

				// Decode b64.
				$_readme_decoded = base64_decode( $_readme_b64 );

				// Init an object.
				$parsedown_obj = new \WGA\Parsedown();

				// Convert markdown to HTML.
				$_readme_html = $parsedown_obj->text( $_readme_decoded );

				// Remove h1.
				if ( 'n' == $_is_h1 ) {
					// Convert all h1's to h2 for better SEO.
					$_readme_remove_h1 = preg_replace( '/<h1[^>]*>([\s\S]*?)<\/h1[^>]*>/', '', $_readme_html );

					// Return it, Safe in PHP 7.0.
					$_return = $_readme_remove_h1;

					// Return!
					return $_return;
				}

				// Convert h1 to h2.
				if ( 'y' == $_ish1toh2 ) {
					// Convert all h1's to h2 for better SEO.
					$_readme_h1toh2 = preg_replace( '/<h1[^>]*>([\s\S]*?)<\/h1[^>]*>/', '<h2>$1</h2>', $_readme_html );

					// Return it, Safe in PHP 7.0.
					$_return = $_readme_h1toh2;

					// Return!
					return $_return;
				}

				// Return it, Safe in PHP 7.0.
				$_return = $_readme_html;

				// Return the data.
				return $_return;
			}

			// Return the data.
			return $wpgapi_response->$_data;
		}

		/**
		 * WP HTTP API GET.
		 *
		 * @param API URL $api_url Remaining API URL from user.
		 * @param Data    $data Data to be retreieved.
		 * @since 0.0.2
		 */
		public function get_api( $api_url, $data ) {
			// Transients are 172 characters only.
			// Get the words only. In array.
			$_api_url_words = explode( '/', $api_url );

			// Save acronym.
			$_api_url_acronym = '';

			// Run through each word of array.
			foreach ( $_api_url_words as $word ) {
				// Uppercase and get last 3 (-3) characters and 3 of them from each word.
				$_api_url_acronym .= strtoupper( substr( $word, -3, 3 ) );
			}

			// Set a transient with API_URL and DATA.
			$transient = 'wga_' . $_api_url_acronym . '_' . $data;

			// Delete trasient for debugging.
			// delete_transient( $transient );

			// Get the value.
			$wga_transient = get_transient( $transient );

			// API URL with $api_url from user.
			$url = 'https://api.github.com/' . $api_url;

			// If no transient then fetch.
			if ( false === $wga_transient || null === $wga_transient || is_wp_error( $wga_transient ) ) {
				// Get the response.
				$response  = wp_safe_remote_get( $url );

				// If not WP_Error.
				if ( ! is_wp_error( $response ) ) {
					// Retreive the body and JSON decode it as an object.
					$wga = json_decode( wp_remote_retrieve_body( $response ) );

					// Set a transient for future.
					set_transient( $transient, $wga, 24 * HOUR_IN_SECONDS );

					// Return the object.
					return $wga;
				} elseif ( true === WP_DEBUG && is_wp_error( $response ) ) {
					return $response->get_error_message();
				} else {
					return '(Unable to fetch at this time!)';
				}
			} else {
				return $wga_transient;
			}
		}
	}

endif;
